### PR TITLE
Add support for New Dooya motor`s (0x4f6E)

### PR DESCRIPTION
**Pull request by burjakremen**
Add support for a new Dooya DT360E-45/20 curtain motor`s (Device ID: 4F6E)

### CODE REVIEW

File: broadlink/__init__.py [hunk 1]
```diff
@@ -86,6 +86,7 @@
     0x2722: (S1C, "S2KIT", "Broadlink"),
     0x4ead: (hysen, "HY02B05H", "Hysen"),
     0x4e4d: (dooya, "DT360E-45/20", "Dooya"),
+    0x4f6e: (dooya_new, "DT360E-45/20", "Dooya"),
```
Comment (felipediel) [inline broadlink/__init__.py:89]: Where did you get the model name from?

**Comment by felipediel** [comment]
@burjakremen Thank you very much for the contribution!

File: broadlink/__init__.py (hunk 1 above)
Comment (burjakremen) [reply broadlink/__init__.py:89]: The sticker on the motor housing and on packaging
dooya_dt360

File: broadlink/cover.py [hunk 2]
```diff
@@ -58,3 +58,55 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+
+    def stop(self) -> int:
+        """Stop the curtain."""
+        return self._send(0xc7, 0xc0, 0x03, 0x00)
+
+    def get_percentage(self) -> int:
+        """Return the position of the curtain."""
+        return self._send(0xca, 0xc0 ,0x06, 0x5d)
+
+    def set_percentage_and_wait(self, new_percentage) -> int:
```
Comment (felipediel) [inline broadlink/cover.py:106]:
```python3
    def set_percentage(self, new_percentage: int) -> int:
```

File: broadlink/cover.py [hunk 3]
```diff
@@ -58,3 +58,55 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+        packet[9] = 0x0b
+        packet[10] = 0x0a
+        packet[15] = magic3
+        packet[16] = magic4
+        response = self.send_packet(0x6a, packet)
+        check_error(response[0x22:0x24])
+        payload = self.decrypt(response[0x38:])
+        return payload[17]
+
+    def open(self) -> int:
```
Comment (felipediel) [inline broadlink/cover.py:90]: What does the integer returned by this function represent?

File: broadlink/__init__.py (hunk 1 above)
Comment (felipediel) [reply broadlink/__init__.py:89]: Beautiful.

File: broadlink/cover.py [hunk 4]
```diff
@@ -58,3 +59,55 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+
+    def stop(self) -> None:
+        """Stop the curtain."""
+        self._send(0xc7, 0xc0, 0x03, 0x00)
+
+    def get_percentage(self) -> int:
+        """Return the position of the curtain."""
+        return self._send(0xca, 0xc0 ,0x06, 0x00)
+
+    def set_percentage_and_wait(self, new_percentage) -> None:
```
Comment (felipediel) [inline broadlink/cover.py:107]: I think this method should be renamed to `set_percentage` because it is non-blocking.

Comment (burjakremen) [reply broadlink/cover.py]: `        return self._send(0xca, 0xc0 ,0x06, 0x00)

    def set_percentage(self, new_percentage) -> None:
        new_percent_hex = struct.pack('<H', 49357+new_percentage)`

Comment (felipediel) [reply broadlink/cover.py]: Does the device wait until the position is reached to send the response?

File: broadlink/cover.py [hunk 5]
```diff
@@ -58,3 +59,55 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+
+class dooya_new(device):
+    """Controls a Dooya curtain motor."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Initialize the controller."""
+        device.__init__(self, *args, **kwargs)
+        self.type = "Dooya DT360E(4F6E)"
+
+    def _send(self, magic1: int, magic2: int, magic3:int, magic4:int) -> int:
```
Comment (felipediel) [inline broadlink/cover.py:71]: Remove the "magic1" and "magic2" parameters.
Rename "magic3" to "command".
Rename "magic4" to "attribute" and set default to 0.

File: broadlink/cover.py [hunk 6]
```diff
@@ -58,3 +59,55 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+        """Send a packet to the device."""
+        packet = bytearray(32)
+        packet[0] = 0x16
+        packet[2] = 0xa5
+        packet[3] = 0xa5
+        packet[4] = 0x5a
+        packet[5] = 0x5a
+        packet[6] = magic1
+        packet[7] = magic2
```
Comment (felipediel) [inline broadlink/cover.py:80]: This magic is called checksum. We can use this:
```python3
        checksum = 0xc0c4 + command + attribute
        packet[6] = checksum & 0xff
        packet[7] = checksum >> 8
```

File: broadlink/cover.py [hunk 7]
```diff
@@ -58,3 +59,55 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+        packet[8] = 0x02
+        packet[9] = 0x0b
+        packet[10] = 0x0a
+        packet[15] = magic3
+        packet[16] = magic4
```
Comment (felipediel) [inline broadlink/cover.py:85]: Rename "magic3" to "command" and "magic4" to "attribute".

File: broadlink/cover.py [hunk 8]
```diff
@@ -58,3 +59,55 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+        response = self.send_packet(0x6a, packet)
+        check_error(response[0x22:0x24])
+        payload = self.decrypt(response[0x38:])
+        return payload[17]
+
+    def open(self) -> None:
+        """Open the curtain."""
+        self._send(0xc5, 0xc0, 0x01, 0x00)
```
Comment (felipediel) [inline broadlink/cover.py:93]:
```python3
self._send(0x01)
```

File: broadlink/cover.py [hunk 9]
```diff
@@ -58,3 +59,55 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+
+    def close(self) -> None:
+        """Close the curtain."""
+        self._send(0xc6, 0xc0, 0x02, 0x00)
```
Comment (felipediel) [inline broadlink/cover.py:97]:
```python3
self._send(0x02)
```

File: broadlink/cover.py [hunk 10]
```diff
@@ -58,3 +59,55 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+
+    def stop(self) -> None:
+        """Stop the curtain."""
+        self._send(0xc7, 0xc0, 0x03, 0x00)
```
Comment (felipediel) [inline broadlink/cover.py:101]:
```python3
self._send(0x03)
```

File: broadlink/cover.py [hunk 11]
```diff
@@ -58,3 +59,55 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+
+    def get_percentage(self) -> int:
+        """Return the position of the curtain."""
+        return self._send(0xca, 0xc0 ,0x06, 0x00)
```
Comment (felipediel) [inline broadlink/cover.py:105]:
```python3
self._send(0x06)
```

File: broadlink/cover.py [hunk 12]
```diff
@@ -58,3 +59,55 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+
+    def set_percentage(self, new_percentage) -> None:
+        new_percent_hex = struct.pack('<H', 49357+new_percentage)
+        magic1 = new_percent_hex[0]
+        magic2 = new_percent_hex[1]
+        magic3 = 0x09
+        magic4 = new_percentage
+        self._send(magic1, magic2, magic3, magic4)
```
Comment (felipediel) [inline broadlink/cover.py:113]:
```python3
self._send(0x09, new_percentage)
```

**Review by felipediel — CHANGES_REQUESTED** [review]

File: broadlink/cover.py [hunk 13]
```diff
@@ -1,5 +1,6 @@
 """Support for covers."""
 import time
+import struct
```
Comment (felipediel) [inline broadlink/cover.py:3]: Remove unused import.

Comment (burjakremen) [reply broadlink/cover.py]: Thanks for your constructive comments. All necessary changes have been made.

File: broadlink/cover.py [hunk 14]
```diff
@@ -58,3 +58,52 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
                 time.sleep(0.2)
                 current = self.get_percentage()
         self.stop()
+
+class dooya_new(device):
```
Comment (felipediel) [inline broadlink/cover.py:62]:
```python3
class dt36b(device):
```

File: broadlink/cover.py [hunk 15]
```diff
@@ -58,3 +58,52 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
...
+    """Controls a Dooya curtain motor."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Initialize the controller."""
+        device.__init__(self, *args, **kwargs)
+        self.type = "Dooya DT360E(4F6E)"
```
Comment (felipediel) [inline broadlink/cover.py:68]:
```python3
        self.type = "DT36B"
```

File: broadlink/__init__.py [hunk 16]
```diff
@@ -5,7 +5,7 @@
 
 from .alarm import S1C
 from .climate import hysen
-from .cover import dooya
+from .cover import dooya, dooya_new
```
Comment (felipediel) [inline broadlink/__init__.py:8]:
```python3
from .cover import dooya, dt36b
```

File: broadlink/__init__.py [hunk 17]
```diff
@@ -92,6 +92,7 @@
     0x2722: (S1C, "S2KIT", "Broadlink"),
     0x4EAD: (hysen, "HY02B05H", "Hysen"),
     0x4E4D: (dooya, "DT360E-45/20", "Dooya"),
+    0x4F6E: (dooya_new, "DT360E-45/20", "Dooya"),
```
Comment (felipediel) [inline broadlink/__init__.py:95]:
```python3
    0x4F6E: (dt36b, "DT360E-45/20", "Dooya"),
```

File: broadlink/cover.py [hunk 18]
```diff
@@ -58,3 +58,52 @@ def set_percentage_and_wait(self, new_percentage: int) -> None:
                 time.sleep(0.2)
                 current = self.get_percentage()
         self.stop()
+
+class dooya_new(device):
+    """Controls a Dooya curtain motor."""
```
Comment (felipediel) [inline broadlink/cover.py:63]:
```python3
"""Controls a Dooya DT360E (type B)."""

**Review by felipediel — APPROVED** [review]
This code looks very good. Latest suggestions...

File: broadlink/cover.py (hunk 18 above)
Comment (burjakremen) [reply broadlink/cover.py]: All done

**Comment by felipediel** [comment]
This code looks really good. I think it is ready to be merged.

My only doubt is about the name of the class. Maybe we should rename dooya to dt1 and this one to dt2? Or use dt360 and dt360b? I don't know. @mjg59 May I have your opinion on this?

**Comment by yuqianma** [comment]
What's status here? I have `0x4f6e` devices too.

**Comment by yuqianma** [comment]
@felipediel Hi, can you merge `0x4f6e`? So we can use it in HA to make things easier. Or is there any simple way to add custom codes instead of forking the whole repo?

**Comment by felipediel** [comment]
Hi. This code is outdated, I will rebase and give it some treatment when I find the time

**Comment by felipediel** [comment]
It's not just a hex code, it needs a PR rebase and Home Assistant integration. It will take some time.

**Comment by felipediel** [comment]
Hi, why did you close, it's a really good addition to the project, could you please reopen it?

**Comment by burjakremen** [comment]
Reopen here: https://github.com/mjg59/python-broadlink/pull/770

**Pull request closed without merging**